src/coreclr/tools/Common/Compiler/ObjectWriter/WasmObjectWriter.cs

```diff
@@ -0,0 +1,244 @@
+            foreach (WasmFuncType signature in functionSignatures)
+            {
+                if (logger.IsVerbose)
+                {
+                    logger.LogMessage($"Emitting function signature: {signature}");
+                }
+
+                int signatureSize = signature.EncodeSize();
+                signature.Encode(buffer.GetSpan(signatureSize));
+                buffer.Advance(signatureSize);
+            }
+
+            return typeSection;
+        }
+
+        private WasmSection WriteFunctionSection(ArrayBuilder<WasmFuncType> allFunctionSignatures, Dictionary<WasmFuncType, int> signatureMap, Logger logger)
+        {
+            WasmSection functionSection = new WasmSection(WasmSectionType.Function, new SectionData(), "function");
+            SectionWriter writer = functionSection.Writer;
+            // Write the number of functions
+            writer.WriteULEB128((ulong)allFunctionSignatures.Count);
+            for (int i = 0; i < allFunctionSignatures.Count; i++)
+            {
+                WasmFuncType signature = allFunctionSignatures[i];
+                writer.WriteULEB128((ulong)signatureMap[signature]);
+            }
+            return functionSection;
+        }
+
+        private void EmitSection(Func<WasmSection> writeFunc, Stream outputFileStream, Logger logger)
+        {
+            WasmSection section = writeFunc();
+            Span<byte> headerBuffer = stackalloc byte[section.HeaderSize];
+
+            section.EncodeHeader(headerBuffer);
+            outputFileStream.Write(headerBuffer);
+            if (logger.IsVerbose)
+            {
+                logger.LogMessage($"Wrote section header of size {headerBuffer.Length} bytes.");
+            }
+
+            section.Data.GetReadStream().CopyTo(outputFileStream);
+            if (logger.IsVerbose)
+            {
+                logger.LogMessage($"Emitted section: {section.Name} of type `{section.Type}` with size {section.Data.Length} bytes.");
+            }
+        }
+
+        protected internal override void UpdateSectionAlignment(int sectionIndex, int alignment) => throw new NotImplementedException();
+        private WasmSection CreateSection(WasmSectionType sectionType, string symbolName, int sectionIndex) => throw new NotImplementedException();
+        private protected override void CreateSection(ObjectNodeSection section, string comdatName, string symbolName, int sectionIndex, Stream sectionStream) => throw new NotImplementedException();
+        private protected override void EmitObjectFile(Stream outputFileStream) => throw new NotImplementedException();
+        private protected override void EmitRelocations(int sectionIndex, List<SymbolicRelocation> relocationList) => throw new NotImplementedException();
+        private protected override void EmitSymbolTable(IDictionary<string, SymbolDefinition> definedSymbols, SortedSet<string> undefinedSymbols) => throw new NotImplementedException();
```

Comment:
Are these overrides needed?

Just asking because the `string` in some of these signatures will clash with #122160 (depending on which PR merges first) because I'm changing this from `string` to `Utf8String`. This and the `internal virtual` change I pointed to above are the only friction points between the two PRs.

I don't mind adapting my PR if this merges first, but if it doesn't merge first, removing these changes (if they're unnecessary) will avoid having to adapt here to the other PR.